kv/kvserver: TestStoreRangeMergeDeadFollowerDuringTxn failed

[(kv/kvserver).TestStoreRangeMergeDeadFollowerDuringTxn failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2402288&tab=buildLog) on [master@b582d9719ebdd0e5456b3ed81265449931b8b03c](https://github.com/cockroachdb/cockroach/commits/b582d9719ebdd0e5456b3ed81265449931b8b03c):

Fatal error:
```
panic: pebble: closed

```

Stack:
```
goroutine 133635 [running]:
github.com/cockroachdb/pebble.(*DB).Apply(0xc001f44000, 0xc0023ca500, 0x7d838d4, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/db.go:533 +0x3f8
github.com/cockroachdb/pebble.(*Batch).Commit(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/batch.go:727
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleBatch).Commit(0xc001536800, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_batch.go:341 +0x5e
github.com/cockroachdb/cockroach/pkg/storage.WriteSyncNoop(0x56e6360, 0xc0020023c0, 0x589e420, 0xc001c542c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine.go:655 +0xd8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.Server.WaitForApplication.func1(0xc00389f500, 0x1, 0xc00389f500)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores_server.go:114 +0x1cb
github.com/cockroachdb/cockroach/pkg/kv/kvserver.Server.execStoreCommand(0xc000afe380, 0x100000001, 0xc003c28930, 0xc003c28950, 0x4249c3)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores_server.go:42 +0x69
github.com/cockroachdb/cockroach/pkg/kv/kvserver.Server.WaitForApplication(0xc000afe380, 0x56e6360, 0xc0020023c0, 0xc0056ce3c0, 0xc000afe380, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores_server.go:88 +0x7d
github.com/cockroachdb/cockroach/pkg/kv/kvserver._PerReplica_WaitForApplication_Handler.func1(0x56e6360, 0xc0020023c0, 0x44064a0, 0xc0056ce3c0, 0x0, 0x0, 0x203000, 0x203000)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/storage_services.pb.go:281 +0x86
github.com/cockroachdb/cockroach/pkg/util/tracing.ServerInterceptor.func1(0x56e6360, 0xc0020023c0, 0x44064a0, 0xc0056ce3c0, 0xc00182b260, 0xc00182b280, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/grpc_interceptor.go:125 +0x579
github.com/cockroachdb/cockroach/pkg/kv/kvserver._PerReplica_WaitForApplication_Handler(0x425dda0, 0xc000afe380, 0x56e6360, 0xc0020023c0, 0xc000f90120, 0xc004b7a210, 0x56e6360, 0xc0020023c0, 0xc000cd8c00, 0xa)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/storage_services.pb.go:283 +0x14b
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0019104e0, 0x57696e0, 0xc000cfc300, 0xc002e1e500, 0xc006f17ef0, 0x7d47f78, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082 +0x4fd
google.golang.org/grpc.(*Server).handleStream(0xc0019104e0, 0x57696e0, 0xc000cfc300, 0xc002e1e500, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0xd25
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc008132d60, 0xc0019104e0, 0x57696e0, 0xc000cfc300, 0xc002e1e500)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:744 +0xa1

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestStoreRangeMergeDeadFollowerDuringTxn

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeMergeDeadFollowerDuringTxn PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeMergeDeadFollowerDuringTxn.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/stores_server.go
+++ b/pkg/kv/kvserver/stores_server.go
@@ -34,12 +34,18 @@ func MakeServer(descriptor *roachpb.NodeDescriptor, stores *Stores) Server {
 	return Server{stores}
 }
 
-func (is Server) execStoreCommand(h StoreRequestHeader, f func(*Store) error) error {
+func (is Server) execStoreCommand(
+	ctx context.Context, h StoreRequestHeader, f func(context.Context, *Store) error,
+) error {
 	store, err := is.stores.GetStore(h.StoreID)
 	if err != nil {
 		return err
 	}
-	return f(store)
+	// NB: we use a task here to prevent errant RPCs that arrive after stopper shutdown from
+	// causing crashes. See #56085 for an example of such a crash.
+	return store.stopper.RunTaskWithErr(ctx, "store command", func(ctx context.Context) error {
+		return f(ctx, store)
+	})
 }
 
 // CollectChecksum implements PerReplicaServer.
@@ -47,8 +53,8 @@ func (is Server) CollectChecksum(
 	ctx context.Context, req *CollectChecksumRequest,
 ) (*CollectChecksumResponse, error) {
 	resp := &CollectChecksumResponse{}
-	err := is.execStoreCommand(req.StoreRequestHeader,
-		func(s *Store) error {
+	err := is.execStoreCommand(ctx, req.StoreRequestHeader,
+		func(ctx context.Context, s *Store) error {
 			r, err := s.GetReplica(req.RangeID)
 			if err != nil {
 				return err
@@ -85,7 +91,7 @@ func (is Server) WaitForApplication(
 	ctx context.Context, req *WaitForApplicationRequest,
 ) (*WaitForApplicationResponse, error) {
 	resp := &WaitForApplicationResponse{}
-	err := is.execStoreCommand(req.StoreRequestHeader, func(s *Store) error {
+	err := is.execStoreCommand(ctx, req.StoreRequestHeader, func(ctx context.Context, s *Store) error {
 		// TODO(benesch): Once Replica changefeeds land, see if we can implement
 		// this request handler without polling.
 		retryOpts := retry.Options{InitialBackoff: 10 * time.Millisecond}
@@ -130,7 +136,7 @@ func (is Server) WaitForReplicaInit(
 	ctx context.Context, req *WaitForReplicaInitRequest,
 ) (*WaitForReplicaInitResponse, error) {
 	resp := &WaitForReplicaInitResponse{}
-	err := is.execStoreCommand(req.StoreRequestHeader, func(s *Store) error {
+	err := is.execStoreCommand(ctx, req.StoreRequestHeader, func(ctx context.Context, s *Store) error {
 		retryOpts := retry.Options{InitialBackoff: 10 * time.Millisecond}
 		for r := retry.StartWithCtx(ctx, retryOpts); r.Next(); {
 			// Long-lived references to replicas are frowned upon, so re-fetch the

